wrapanapi/containers/providers/rhopenshift.py

```diff
@@ -0,0 +1,892 @@
+        """Returns list of projects"""
+        entities = []
+        entities_j = self.o_api.get('project')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = Project(self, meta['name'])
+            entities.append(entity)
+        return entities
+
+    def list_template(self):
+        """Returns list of templates"""
+        entities = []
+        entities_j = self.o_api.get('template')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = Template(self, meta['name'], meta['namespace'])
+            entities.append(entity)
+        return entities
+
+    def list_docker_image(self):
+        """Returns list of images (Docker registry only)"""
+        entities = []
+        entities_j = self.o_api.get('image')[1]['items']
+        for entity_j in entities_j:
+            if 'dockerImageReference' not in entity_j:
+                continue
+            _, name, image_id, _ = Image.parse_docker_image_info(entity_j['dockerImageReference'])
+            entities.append(Image(self, name, image_id))
+        return entities
+
+    def list_deployment_config(self):
+        """Returns list of deployment configs"""
+        entities = []
+        entities_j = self.o_api.get('deploymentconfig')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = DeploymentConfig(self, meta['name'], meta['namespace'])
+            entities.append(entity)
+        return entities
+
+    def deploy_template(self, template, tags=None, password='smartvm', **kwargs):
+        """Deploy a VM from a template
+
+        Args:
+            template: (str) The name of the template to deploy
+            tags: (dict) dict with tags if some tag isn't passed it is set to 'latest'
+            password: this password will be set as default everywhere
+        Returns: dict with parameters necessary for appliance setup or None if deployment failed
+        """
+        self.logger.info("starting template %s deployment", template)
+        self.wait_template_exist(namespace=self.default_namespace, name=template)
+
+        if not self.base_url:
+            raise ValueError("base url isn't provided")
+
+        prepared_tags = {key: 'latest' for key in self.template_tags}
+        if tags:
+            not_found_tags = [tag for tag in tags.keys() if tag not in self.stream_tags]
+            if not_found_tags:
+                raise ValueError("Some passed tags {t} don't exist".format(t=not_found_tags))
+            for tag, value in tags.items():
+                prepared_tags[self.stream2template_tags_mapping[tag]] = value
+
+        # create project
+        # assuming this is cfme installation and generating project name
+        proj_id = "".join(choice(string.digits + string.lowercase) for _ in range(6))
+        proj_name = "{t}-project-{pid}".format(pid=proj_id, t=template)
```

Comment:
consistent formatting keys would make this a bit more readable - `pid` makes me think `process ID`, but `proj` used below is more descriptive